Java: Connecting Errors with Spans (#2818)
getsentry/sentry-docs: src/platforms/java/common/performance/index.mdx

    }
}
```

### Connecting Errors with Spans

Sentry errors can be linked with transactions and spans.

Errors reported to Sentry while transaction or span is running are linked automatically:

> ```java

Reviewer: should we add kotlin snippets as well?

Author: 👍

ISpan span = Sentry.startTransaction("task");
try {
    processItem();
} catch (Exception e) {
    Sentry.captureException(e);
} finally {
    span.finish();

> }

Reviewer: should we split these 2 code snippets into the include folder? they are gonna be reused by Android and right now they are under `platforms/java/common/performance/index.mdx`.

otherwise, every fix should be copied over 2 MDX files

```

In cases where the decision if thrown exception should be reported to Sentry happens later in the future after span is already finished, to connect error with a span you must:

1. Link exception with a Span by calling `setThrowable` method:

```java
ISpan span = Sentry.getSpan();
try {
    processItem();
} catch (Exception e) {
    span.setThrowable(e);
    throw e;
} finally {
    span.finish();

> }

Reviewer: take into consideration

https://github.com/getsentry/sentry-docs/pull/2818/files#r553794887
https://github.com/getsentry/sentry-docs/pull/2818/files#r553834912

```

2. Retrieve span context associated with exception when reporting an error with `IHub#getSpan`:

```java
SentryEvent event = new SentryEvent(throwable);

> SpanContext spanContext = HubAdapter.getInstance().getSpan(ex);

Reviewer 2: If we're going for `HubAdapter` it means we need to rethink our API here.

Is `getSpan` supposed to be on the main `Sentry` class? 

Better:

Can we delete this section and simply have that in the docs? When we capture an exception we lookup for a span in the weakref? don't think this should be a user concern

Author: This means that we would need to move `WeakHashMap<Throwable, SpanContext> throwableToSpanContext` from `Hub` to `SentryClient` so that it can be applied for `captureEvent`, `captureEvent` and `captureException`. Which is fine for me but I am not sure if it does not violate any rules regarding what can be inside `SentryClient` cc @marandaneto

Reviewer: `HubAdapter` is only meant for easing unit tests, not to be used externally, so definitely we'd need to rethink the API.

https://develop.sentry.dev/sdk/unified-api/
Hub should manage the state and `throwableToSpanContext` looks like managing the state.
Client is responsible for building the events only, no state at all afaik.

I didn't have the chance to read/test the whole transaction feature, not sure if I'm already able to give a better option, but working on it.

Reviewer: > Sorry but I don't follow:
> Why does it need to go to the client? The API entry point is `Sentry.captureX` which goes through the Hub. We don't want to have `getSpan` there? Is this method not in the spec?
> 
> Lets just remove this from the docs for now, this was created to support our Spring integration and doesn't need to be documented yet.

yeah I also don't follow, don't really understand why we need that, I'm blind here.

> 2. Retrieve span context associated with exception when reporting an error with `IHub#getSpan`:
> 
> ```java
> SentryEvent event = new SentryEvent(throwable);
> SpanContext spanContext = HubAdapter.getInstance().getSpan(ex);
> if (spanContext != null) {
>   event.getContexts().setTrace(spanContext);
> }
> hub.captureEvent(event);
> ```

Reviewer 2: ```suggestion
```

Reviewer 2: Ignore this, lets discuss how to add this OOTB first

    }
}
```

### Connecting Errors with Spans

Sentry errors can be linked with transactions and spans.

Errors reported to Sentry while transaction or span is running are linked automatically:

```java {tabTitle: Java}
ISpan span = Sentry.startTransaction("task");
try {
    processItem();
} catch (Exception e) {
    Sentry.captureException(e);
} finally {
    span.finish();
}
```

```kotlin {tabTitle: Kotlin}
val span = Sentry.startTransaction("task")
try {
    processItem()
} catch (e: Exception) {
    Sentry.captureException(e)
} finally {
    span.finish()
}
```

> In cases where the decision if thrown exception should be reported to Sentry happens later in the future after span is already finished, to connect error with a span you must link the exception with a Span by calling `setThrowable` method:

Reviewer 3: This is confusing. I think this means (and feel free to modify if I've got this wrong):
```suggestion
In future, if a thrown exception should be reported to Sentry after the span has already finished, you must link the exception with the span by calling `setThrowable` method to connect the error:
```

Author: I am not sure if "In future" here is understandable for everyone. Maybe we can just leave "If a thrown exception ..." ?

```java {tabTitle: Java}
ISpan span = Sentry.getSpan();
try {
    processItem();
} catch (Exception e) {
    span.setThrowable(e);
    throw e;
} finally {
    span.finish();
}
```

```kotlin {tabTitle: Kotlin}
val span = Sentry.getSpan()
try {
    processItem()
} catch (e: Exception) {
    span.setThrowable(e)
    throw e
} finally {
    span.finish()
}
```

> Please note, that in Spring Boot integration, errors are linked with spans automatically and there is no need to follow these steps.

Reviewer 3: Move this up earlier in the page so Spring Boot users don't get to the end, only to learn they didn't need to read this far :).
```suggestion
<Note>

In our Spring Boot integration, errors are linked with spans automatically.

</Note>
```

> Please note, that in Spring Boot integration, errors are linked with spans automatically and there is no need to follow these steps.

Reviewer 2: Isn't it "on our Spring Integration"?
Is this exclusive to Spring Boot? We've got customers asking for this that use Spring but not Spring Boot.

Author: Correct, updated.
